### ISSUE TITLE
An unstable unit test: BrokerOuterAPITest.test_needRegister_timeout

### ISSUE THREAD

**Issue by zhouxinyu** [apache/rocketmq#6654]
Failed detail:
>   BrokerOuterAPITest.test_needRegister_timeout:141 expected:<2> but was:<3>

**Comment by ShadowySpirits** [comment]
> @zhouxinyu can I work on this ?

Welcome! Looking forward to your pr.

**Comment by KitherRunner** [comment]
Modify return type：List<Boolean> --> Boolean[]
  org.apache.rocketmq.broker.out.BrokerOuterAPI#needRegister
  return changedList.toArray(new Boolean[0])？

**Comment by Abhijeetmishr** [comment]
@KitherRunner Why we should change the return type of the method from List<Boolean> to Array of Boolean ?

**Comment by KitherRunner** [comment]
> @KitherRunner Why we should change the return type of the method from List to Array of Boolean ?

When CountDownLatch times out and continues to run, the executed tasks will not be canceled, and they can still add elements to the List

**Comment by Abhijeetmishr** [comment]
> > @KitherRunner Why we should change the return type of the method from List to Array of Boolean ?
> 
> When CountDownLatch times out and continues to run, the executed tasks will not be canceled, and they can still run and add elements to the List

No, when a Countdown Latch times out and continues to run, the executed tasks will not be canceled, but they will not be able to add elements to the List.
This is what I found out I can be wrong.

**Comment by KitherRunner** [comment]
> > > @KitherRunner Why we should change the return type of the method from List to Array of Boolean ?
> > 
> > 
> > When CountDownLatch times out and continues to run, the executed tasks will not be canceled, and they can still run and add elements to the List
> 
> No, when a Countdown Latch times out and continues to run, the executed tasks will not be canceled, but they will not be able to add elements to the List. This is what I found out I can be wrong. I think we have to look at it from different angle any suggestion @zhouxinyu.

The later the return value List is used, the more obvious the effect. Tasks that have timed out can still be added to the List with elements

**Comment by KitherRunner** [comment]
The later the return value List is used, the more obvious the effect. Tasks that have timed out can still be added to the List with elements. Just ensure that the return value remains the same after the timeout, and there's no problem

**Comment by KitherRunner** [comment]
> > > @KitherRunner Why we should change the return type of the method from List to Array of Boolean ?
> > 
> > 
> > When CountDownLatch times out and continues to run, the executed tasks will not be canceled, and they can still run and add elements to the List
> 
> No, when a Countdown Latch times out and continues to run, the executed tasks will not be canceled, but they will not be able to add elements to the List. This is what I found out I can be wrong. I think we have to look at it from different angle any suggestion @zhouxinyu.

The later the return value List is used, the more obvious the effect. Tasks that have timed out can still be added to the List with elements. Just ensure that the return value remains the same after the timeout, and there's no problem

**Comment by Abhijeetmishr** [comment]
> > > > @KitherRunner Why we should change the return type of the method from List to Array of Boolean ?
> > > 
> > > 
> > > When CountDownLatch times out and continues to run, the executed tasks will not be canceled, and they can still run and add elements to the List
> > 
> > 
> > No, when a Countdown Latch times out and continues to run, the executed tasks will not be canceled, but they will not be able to add elements to the List. This is what I found out I can be wrong. I think we have to look at it from different angle any suggestion @zhouxinyu.
> 
> The later the return value List is used, the more obvious the effect. Tasks that have timed out can still be added to the List with elements. Just ensure that the return value remains the same after the timeout, and there's no problem

okay thank you.

**Comment by abhishku** [comment]
Hi, Its my first time here :) May I start working on this ?

**Comment by ashutosh-roy** [comment]
Hi @zhouxinyu , Couldn't reproduce this scenario in my local. This test is passing.

<img width="370" alt="image" src="https://github.com/apache/rocketmq/assets/44959785/f8a9d058-c2b3-4f84-bf22-15fe4cbccb64">

**Comment by arsenalzp** [comment]
Any progress with this issue?

**Comment by akagra** [comment]
Hello I want to contribute towards this issue. Can you guys assign this to me?

**Comment by akagra** [comment]
@zhouxinyu Can I start work on this issue I am glads to work on this

**Comment by sam6134** [comment]
Instead of returning the original list, why not we return a copy of the same list? I think that should cause the flakiness to go. Seems the even after the latch countdown threads still can modify the original list. If we copy to a different list after countDown I think that should fix it
